File: frontend/src/main/java/app/programmatic/ui/reporting/segments/SegmentsReportService.java
package app.programmatic.ui.reporting.segments;

import app.programmatic.ui.reporting.model.ReportFormat;
import app.programmatic.ui.reporting.validation.ValidateReportParameters;
import app.programmatic.ui.reporting.view.ReportMeta;

public interface SegmentsReportService {

    byte[] runReportSegments(@ValidateReportParameters SegmentsReportParameters parameters, ReportFormat format);

    ReportMeta getReportMeta(SegmentsReportParameters parameters);
}

File: ui/src/app/advertiser/services/segments_report.service.ts
import {Injectable} from '@angular/core';
import {HttpClient} from '@angular/common/http';
import {Router} from '@angular/router';
import {CommonService} from '../../common/services/common.service';

@Injectable()
export class SegmentsReportService extends CommonService {

  constructor(public router: Router, public http: HttpClient) {
    super(router, http);
  }

  getReportMeta(parameters: SegmentsReportParameters): Promise<ReportMeta> {
    return this.httpPut(this.api.report.meta.segments, parameters);
  }

  generateReport(parameters: SegmentsReportParameters): Promise<any> {
    return this.httpPut(this.api.report.run.segments, parameters, {format: 'JSON'});
  }

  downloadReport(parameters: SegmentsReportParameters, format: string): Promise<any> {
    return this.httpDownload('PUT', this.api.report.run.segments, parameters, {format});
  }
}
